### DISCUSSION TITLE
Support for InfluxDB 2.x>

### DISCUSSION THREAD

**Discussion by luisfocosta** [David00/rpi-power-monitor#48] [Ideas]
I have already a running a Influxdb instance 2.1.1 and I assume I can configure a remote instance - in config.py
However what is the possibility to support a 2.x version?

**Comment by David00** [comment]
Hi Luis, @BCNelson has already begun working on adding support for Influx 2.x in PR #47.  I haven't approved it yet simply because I haven't tested it, but you're welcome to try it out.  

You can pull BCNelson's fork with the changes into a separate directory and test it on your Pi from there. Just make sure to stop the original project from running with `sudo systemctl stop power-monitor.service` (assuming you've setup the service file).  Example:

```
cd ~/
mkdir rpi-power-monitor-influx2
cd rpi-power-monitor-influx2
git clone https://github.com/BCNelson/rpi-power-monitor.git
cd rpi-power-monitor
git checkout feature/influxdb2
```

So, from the `~/rpi-power-monitor-influx2/rpi-power-monitor/` directory, you can edit config.py to add your influxDB settings.  You'll also need to install the additional influxdb library with `pip install -r ~/rpi-power-monitor-influx2/rpi-power-monitor/requrements.txt`.  Then, you should be able to start the software with `python3 power-monitor.py` from inside the `~/rpi-power-monitor-influx2/rpi-power-monitor/` directory.

**Reply by BCNelson** [reply]
My fork also does not setup the database. You will need to do that on your own or give it a token with the necessary permissions and uncomment some code. Although i have not tested that.

**Reply by luisfocosta** [reply]
Thanks! I will give it a shot

**Comment by acabrera2012** [comment]
Hello. Has anyone had success with version 2.x? I have a client that would like to use a more recent version of InfluxDB and they are concerned about using 1.8.

**Reply by David00** [reply]
This PR is now out of date as it was built for the previous version of the RPI Power Monitor software (v0.2.x).  Quite a few things have changed in the current version (v0.3.x) which are breaking when compared to v0.2.x.  

I am currently working on v0.4.x, and I'll look at including support for a remote Influx 2.X database (either Influx Cloud or a self-hosted remote instance) in this release. There's a bit more that goes along with adding support for Influx v2, like new dashboards, documentation, backups, etc.

**Reply by acabrera2012** [reply]
Thanks, David. That's awesome. I'll check the repo regularly to see how v0.4.x is coming along. I appreciate it.

**Reply by David00** [reply]
Sounds good! Would you like to help test any early release candidates?

**Reply by acabrera2012** [reply]
Yes. I have a working system on v0.3 on a Pi 4 and would love to give the new stuff a try and even do it on a Pi 5, if possible.

**Reply by David00** [reply]
Great! Also planning to release an updated custom OS image that supports Pi 5 too!

**Reply by mholzma** [reply]
I am also up for testing on my Pi 3 going to Influx 2.x DB on another server actually. As an interim approach, and to test stuff out, I created a plugin for the power monitor that writes out data to Influx 2.x every second. My use case for this is a bit different as I am just monitoring every time the circuit turns on/off.

**Reply by acabrera2012** [reply]
Sounds good. I was actually hoping that the final solution would allow for specifying the interval for sending the InfluxDB 2.x or 3.x data to the remote server, like hourly, to reduce processing.

**Comment by acabrera2012** [comment]
Is there anything I can help with on this, David? Research? Testing? Other? ......

**Reply by David00** [reply]
Sure! Also, tagging @mholzma.  

I just pushed the `develop-v0.4.0` branch that has my initial changes to support InfluxDB v2.  I've briefly tested it with my Influx Cloud account, but have not tested it with a locally hosted Influx v2 server, or a real power monitor that's measuring power.

(Side note: it looks like Influx v2 requires a 64-bit OS, and my custom Pi OS image is still 32-bit)

Assuming you are on my custom OS image, you can checkout the new branch with the following commands:

(The sed command just updates the local git config file so that it can track all branches from the repo - no harm in running it anyways if you're not sure whether you need it)

```bash
cd ~/rpi_power_monitor
sed -i 's|+refs/tags/v0.3.*:refs/tags/v0.3.*|+refs/heads/*:refs/remotes/origin/*|g' .git/config
git fetch
git checkout develop-v0.4.0
pip install -r requirements.txt
pip install .
```

Next, you'll have to update your config.toml file with a few new settings.  Here are the full sections for `[database]` and `[database.influx_v2]`.  Note that the host/port/username/password/database_name become irrelevant when `influx_version = 2`.

```toml

[database]
enabled = true
host = "localhost"
port = 8086
username = "root"
password = "password"
database_name = "power_monitor"
influx_version = 2

[database.influx_v2]
bucket = "power_monitor"
org = "<your Influx Cloud email or custom defined org>"
token = "<an API token with at least r/w permissions for all buckets and all tasks>"
url = "<your unique Influx Cloud or self-hosted InfluxDB v2 server URL>"
```

Influx API tokens are setup under the "Load Data" section of the cloud UI.
The URL is just the base Influx URL, like this for example (yours may be different): `https://us-east-1-1.aws.cloud2.influxdata.com`

The bucket will be created automatically if it doesn't exist, so once you get the settings dialed in, you should be good to launch the power monitor. I'd recommend doing it from the terminal with the verbose flag so you can see some details of the startup, like this:

```bash
cd ~/rpi_power_monitor
python3 rpi_power_monitor/power_monitor.py -v
```

Some things still to do in the code:
* create the Influx v2 Tasks for automatic downsampling
* Ensure a bucket exists for the downsampled data; create it if not.
* create retention rules for both buckets and make sure they exist
* write unit tests to handle config parsing, RPiPowerMonitor class instantiation, and general operation

Other related tasks:
* create new Grafana dashboards to display the data

**Reply by acabrera2012** [reply]
Ok. I'll give it a shot. Thanks

**Comment by acabrera2012** [comment]
Followed the instructions but ran into a small hurdle. There's probably something silly happening with my paths but I get this error when I try to run using:

`python3 rpi_power_monitor/power_monitor.py -v`

Error:

`Traceback (most recent call last):
  File "/home/pi/rpi_power_monitor/rpi_power_monitor/power_monitor.py", line 1601, in <module>
    from rpi_power_monitor.plugin_handler import Plugin
ModuleNotFoundError: No module named 'rpi_power_monitor.plugin_handler'
`

I don't want to add sys.path to power_monitor.py because I'm sure there's some other explanation for the error. Let me know if this is something simple or if I should just keep researching.

**Reply by David00** [reply]
Thanks for testing. I forgot a step in the setup instructions!

Can you do the following?

```bash
cd ~/rpi_power_monitor
pip install .
```
(Note the space + period at the end of pip install!)

This should let it work.  

Also, I am aware of the issue when trying to use influx version 1, and the warning message from the influx API when using influx version 2 after starting the code for the 2nd time, which looks like this:


> WARNING :   INFLUX - there was a problem when creating bucket. Message:
WARNING : (403)
Reason: Forbidden
... truncated ...
HTTP response body: {
        "code": "forbidden",
        "message": "creating bucket would exceed quota"
}

This message is ok and you can still use Influx version 2.  However, for Influx v1, I'll have to fix that because it is broken right now in this development version.

**Reply by acabrera2012** [reply]
Ah. Makes sense on the pip install. I'll give it a shot.....

**Reply by acabrera2012** [reply]
That worked! I don't have the CTs connected to any live wires but it's still logging idle noise values to my InfluxDB Cloud account. It has been for hours now. So far so good. I'll keep it running overnight just to confirm everything is stable. I'll be in touch....

**Comment by mholzma** [comment]
I have it up and running and right now it is storing data on a personal
influx2 server. I'll see how it runs for a bit and let you know if I
see/hear anything. Have to go rebuild a bunch of dashboards but it is
pulling from one of the CTs.

- Mike

**Comment by mholzma** [comment]
I also wrote a very quick/simple downsampling task.. I'll have to make sure
it is correct.

**Comment by acabrera2012** [comment]
Ran fairly well for the past 16 hours but stopped in the middle of the night for some reason. I thought I was running it in the background but using a Terminal window instead of running it as a service. The Terminal window probably timed out.

Anyway, not bad for a first test.

What's next?

**Reply by David00** [reply]
If you were connected remotely via SSH, it's possible the SSH connection was interrupted, which would also have killed the power monitor process you started. If this was a local terminal, maybe a bug.

**Reply by acabrera2012** [reply]
Yeah. No real concern there about the Terminal. But, I wasn't able to run it as a service at the time even though it gave no error. I didn't see any data flowing. I forgot to check the logs before I shut it down, though, so that's my fault.

**Comment by acabrera2012** [comment]
Going to try with live connection readings now just to confirm.....

**Comment by acabrera2012** [comment]
Oddly only showing data for one of the CTs but I'm not convinced it's an issue with the new code....

**Comment by mholzma** [comment]
It has been running fine in my instance. I only have one CT connected but
the other 5 are recording data, just zeros. I was wondering how frequently
it sends data to the InfluxDB. Does it batch up some of the data so if
sending to a remote server it sends a group of points rather than a point
each time?

I haven't read through all the code to look at that, but it could be useful
to have a configuration setting for managing the frequency for when you
don't need full realtime data, in case you want to batch every second, or
every minute, etc. This could be a feature request for offboard and cloud
servers to reduce the frequency of transactions.

- Mike

**Reply by acabrera2012** [reply]
It looks like it takes a number of samples and then sends the data to the DB. Not sure if that can be changed to a time interval, like every hour, but that would be awesome.

**Reply by David00** [reply]
Glad to hear yours has been running fine, @mholzma! 

The database writes are batched, albeit a fairly small batch.  A single batch consists of 25 measurements, and once there's 25, the batch is written to Influx. This approach balances the overhead in frequent DB writes with the real-time intent of the project. I wouldn't really want to introduce a time-based approach in the core project, but I could see this being a useful plugin instead.  A plugin could do its own batching and writes to Influx.

**Comment by mholzma** [comment]
I could imagine that given the sample rate and the data there may be in
memory constraints as well as risk of data loss. But could still be handy.
Might be able to do it within the db also.

I offloaded Grafana because my Pi3 was lagging a bit.

**Comment by acabrera2012** [comment]
So, what's next? Anything I can help with @David00 or @mholzma that won't slow down your current progress?

**Reply by David00** [reply]
Sorry, just seeing this! There are some additional things I still need to find time to do, like new dashboards and additional testing.  The dashboards would be a great thing to help out with if you're good with Grafana!

**Reply by acabrera2012** [reply]
Unfortunately, I have zero experience with Grafana at all. I could certainly learn it, from scratch, but I think that would just slow everything down.

Is there any other brainstorming, testing, or research I can help you with?

**Comment by acabrera2012** [comment]
Happy holidays, everyone!

Fyi. I'm going to continue to run the 0.4 code to see if there are any hourly or overall InfluxDB Cloud limitations on the Free Tier.

If there's anything else I can do or test in the mean time, let me know.

Cheers.

**Reply by David00** [reply]
Happy Holidays!

Sounds good. The only impacting limitation to the free tier is the 30 day retention. The read/write limits will not be hit with the power monitor.

**Comment by mholzma** [comment]
I’m okay with Grafana if here is something you want me to work on.. I would
need to pull down the default ones I think, since I don’t use them.

- Mike

**Comment by acabrera2012** [comment]
Fyi. So far, everything appears to be running smoothly with continuous data pushes from Power Monitor.

InfluxDB Cloud does have some other limits for the free tier as follows:

`
Data-in: Rate of 5 MB per 5 minutes (average of 17 kb/s)

Read: Rate of 300 MB per 5 minutes (average of 1000 kb/s)

Storage: 30 days of data retention
`

I didn't see an easy way, within InfluxDB Cloud, to tell if Power Monitor exceeds the Data-in or Read limits. I didn't see any warnings or errors either.

I don't think we're exceeding anything but it is possible that InfluxDB Cloud quietly truncates or discards excess data.

**Comment by acabrera2012** [comment]
Hello, guys. I hope all is well. Anything new and interesting to report on your efforts?

**Reply by David00** [reply]
Work is still underway - there's a lot that goes into building InfluxDB Tasks (even manually in the Influx UI), but then getting them to be created automatically by the power monitor software is another layer of complexity. I'm nearly finished with the automatic task creation, though.

Ultimately, the Influx version 2 support will be rolled out with v0.4.0 release along with some other features.

**Reply by acabrera2012** [reply]
Excellent! Thanks for the update.

**Comment by David00** [comment]
Hello! I've just added a little bit more functionality to the Influx v2 support for v0.4.0.  The power monitor will now create the Influx DB Tasks which handle the downsampling of high resolution data, and the calculation of energy from the high resolution data.

If you're already running the `develop-v0.4.0` branch, you can get the updates with the following commands:

```bash
cd ~/rpi_power_monitor
git pull
pip install . --no-cache
```

Then, restart your power monitor service with `sudo systemctl restart power-monitor.service`

If you're not running the `develop-v0.4.0` branch, you can switch to it by following the instructions in [my comment above](https://github.com/David00/rpi-power-monitor/discussions/48#discussioncomment-11586876).

Now that the 5-minute data will be populated we can build suitable Grafana dashboards.  The measurements and fields are very similar to the existing InfluxDB V1 measurements and fields. I'll have to write docs for Influx V2 structure (see Influx V1 structure [here](https://david00.github.io/rpi-power-monitor/docs/v0.3.0/database-info.html#structure)), but for the time being, the data is explore-able in the Influx V2 UI:

![image](https://github.com/user-attachments/assets/b7087031-c782-4234-9b10-96ad4e62d2d2)

**Reply by acabrera2012** [reply]
That's great, David. Can you tell me how this impacts the InfluxDB Cloud setup? That's what I've been testing on my end.

**Reply by David00** [reply]
When you pull this latest update and restart the power monitor application, it will check to see if the two required tasks exist. If they don't, it will create them. I am not aware of any limitations when it comes to Tasks on InfluxDB Cloud - only read/write data limits.

While the added tasks will use a little bit of the read/write quota, there's not enough data to exceed the limitations of the free tier.

**Reply by acabrera2012** [reply]
Ok. I'll try it out. Thanks

**Reply by acabrera2012** [reply]
@David00, I can see the latest change adds "_downsampled" to the bucket name in InfluxDB Cloud. But, the Measurements drop-down right below that now shows nothing. Can you confirm that the InfluxDB Cloud screen still works for you with the latest downsampling change?

**Reply by David00** [reply]
Hi @acabrera2012, thanks for bringing this up. I just checked mine and it's behaving the same way. At first glance it appears to be related to the automated Influx Task creation (and the ID that was used). I'll look more into it.

**Reply by acabrera2012** [reply]
Sounds good. I appreciate it.

**Reply by David00** [reply]
This should be fixed now! First, go into your InfluxDB dashboard and delete the existing downsample tasks.

Then, on your Pi, do the following to pull and install the fixes from the repo:

```bash
cd ~/rpi_power_monitor
git pull
pip install .
```

Then, stop the power monitor service, and start it manually from the CLI in verbose mode (still from `~/rpi_power_monitor`):


```bash
sudo systemctl stop power-monitor.service
python3 rpi_power_monitor/power_monitor.py -v
```

In the output, you should see the following:

>DEBUG :   Influx v2 - created task powermon_high_res_downsample
DEBUG :   Influx v2 - created task powermon_calculate_5m_energy
DEBUG : Influx version 2 database checks/setup complete.

Then, you can stop the power monitor in the terminal and restart the background service:

```bash
sudo systemctl start power-monitor.service
```

**Reply by acabrera2012** [reply]
Hello, David. Can you clarify the part about deleting existing downsample tasks within the InfluxDB Cloud dashboard? I do see data being pushed under the normal bucket name but the one with "downsampled" at the end does not show any options in the Select Measurement drop-down.

**Reply by David00** [reply]
Sure!  

Go into the Tasks section of your cloud account, and remove the two existing tasks:

![image](https://github.com/user-attachments/assets/41b9f275-bd39-4013-950b-0198e84c3098)

They will be recreated (properly this time) when you pull the latest update and restart the power monitor.

(For context, these Tasks are what run automatically every 5 minutes and calculate the 5 minute downsamples from the `power_monitor` bucket)

**Reply by acabrera2012** [reply]
I see now. In the InfluxDB Cloud dashboard, the Tasks option is not available on the left-hand menu. But, I am able to see and delete the "Buckets" as you can see in this screenshot. Maybe the terminology is different for the cloud vs the self-hosted version?

<img width="1722" alt="Screenshot" src="https://github.com/user-attachments/assets/1280ae5c-cd8d-4dc2-a51a-37eed37f0652" />

**Reply by David00** [reply]
Hmm, perhaps your user doesn't have permissions to see/modify tasks? My screenshot was from InfluxDB cloud ;)

If you click on your "Dev" dropdown at the top you should have a section for members (if you're an admin) and you can see what role your account has. My account's role is "owner" for comparison:

![image](https://github.com/user-attachments/assets/f97c0b38-f8eb-41b2-b6d1-706d3a33d7e7)

**Reply by acabrera2012** [reply]
I am listed as the only Member and my Role is Owner. So, I'm not sure what is different between our two setups. I must be missing something obvious here. I think there must be different Cloud versions. How long ago did you create your Cloud account?

**Reply by acabrera2012** [reply]
My URL shows cloud2: [https://us-east-1-1.aws.cloud2.influxdata.com/orgs/](url)

**Reply by acabrera2012** [reply]
Ah. Someone else asked about the missing menu items in their community forum and here was the response from InfluxDB:

"Those features aren’t supported in InfluxDB 3.0 (previously known as IOx). They use the Flux language which is only supported on the Cloud2 platform (the TSM backed version). We will be adding functionality over time, but for the time being we’re recommending that users looking for Tasks and Alerts functionality use AWS Lambda or some equivalent paired with an InfluxDB3 client library (the Python one works well). 

For the other UI features, we have a limited UI for version 3.0 for the moment. Our focus is mostly on the core of the database. For users looking for a richer UI, they can pair InfluxDB 3.0 with Grafana, Apache Superset, Tableau, and (soon) PowerBI.

Grafana and Superset also have alerting features that can be used with InfluxDB 3.0."

**Reply by acabrera2012** [reply]
Is there another way to clear out those tasks behind the scenes?

**Reply by David00** [reply]
(Had to catch a flight home, hence the abrupt stop earlier!)

That is really weird. I don't quite get why Influx would have taken away the front end for Task management in InfluxDB 2.  My account is almost 3 years old so maybe that has something to do with it.

Anyways, I put together this small script to go through your existing tasks and give you the option to delete each one.  I tested it against my Influx cloud account and it works fine.  I don't necessarily want to put this in the project repo because now that I've fixed the task creation issue, future users shouldn't have this issue.

If you paste the following script into a new file, let's say `delete_tasks.py` and run it, it should work for you.  The file has to be in the ~/rpi_power_monitor/rpi_power_monitor directory due to relative import requirements.

Here are some quick setup steps:

```bash
sudo systemctl stop power-monitor.service
nano ~/rpi_power_monitor/rpi_power_monitor/delete_tasks.py
```

And then paste in this script, (and save, and exit nano of course):

```python3
import influxdb_client
from power_monitor import RPiPowerMonitor, logger, ch

if __name__ == '__main__':
    # Use RPiPowerMonitor to load config
    rpm = RPiPowerMonitor()
    v2_config = rpm.config.get('database').get('influx_v2')

    org = v2_config.get('org')
    token = v2_config.get('token')
    url = v2_config.get('url')      

    client = influxdb_client.InfluxDBClient(
            url=url,
            token=token,
            org=org
        )

    task_api = client.tasks_api()
    existing_tasks = task_api.find_tasks()

    print(f"Looking for tasks in your InfluxDB Org named '{org}' at '{url}'...\n")
    for task in existing_tasks:
        delete = input(f"Found InfluxDB Task named '{task.name}'.  Delete? [y|n]: ")
        if delete.lower() == 'y':
            task_api.delete_task(task.id)
            print(f"Deleted task {task.name}")   

    rpm._cleanup()
```
Then run it...

```bash
cd ~/rpi_power_monitor/rpi_power_monitor
python3 delete_tasks.py 
```

(Go through the script prompts, entering `y` for the powermon tasks that it asks about)

Then, restart the power monitor in verbose mode as mentioned in my earlier comment: 

```bash
cd ~/rpi_power_monitor
python3 rpi_power_monitor/power_monitor.py -v
```

And you should see the messages in the log output about creating the tasks again.

Assuming that's all good, you can stop the power monitor with Ctrl-C, and then don't forget to restart it in the background with: 

```bash
sudo systemctl start power-monitor.service
```

**Reply by acabrera2012** [reply]
Thanks, David. I'll test it out.

**Reply by acabrera2012** [reply]
All tests look good on my end. Thank, again!

**Reply by David00** [reply]
> All tests look good on my end. Thank, again!

Thanks for the follow up!  Just FYI, this support for InfluxDB 2.x will be included with the v0.4.0 release.  Thanks for helping to test out some early features!

**Reply by acabrera2012** [reply]
Got it. Thanks
